service/controller/v11/resource/pod/delete.go

```diff
@@ -38,80 +41,201 @@ func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+			resourcecanceledcontext.SetCanceled(ctx)
+			r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource for custom object")
+
+			return nil
+		}
+	}
 
-	// TODO drain guest cluster node and only remove the finalizer as soon as the
-	// guest cluster node associated with the reconciled host cluster pod is
-	// drained.
+	if !forcePodCleanup(currentPod) {
+		r.logger.LogCtx(ctx, "debug", "found the current version of the reconciled pod in the Kubernetes API")
 
-	// TODO go ahead and do not block once the draining is initialized. Using
-	// finalizers on the pods the delete event will be replayed and we can check
-	// if the draining completed on the next reconciliation loop.
+		n := currentPod.GetNamespace()
+		p := currentPod.GetName()
+		o := metav1.GetOptions{}
 
-	// Here we remove the 'draining-nodes' finalizer from the reconciled pod, if
-	// any. This frees the garbage collection lock in the Kubernetes API and makes
-	// the pod vanish.
-	var podToDelete *corev1.Pod
-	{
-		var changed bool
-		var newFinalizers []string
+		nodeConfig, err := r.g8sClient.CoreV1alpha1().NodeConfigs(n).Get(p, o)
+		if apierrors.IsNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find node config for guest cluster node")
 
-		for _, f := range currentPod.GetFinalizers() {
-			if f == key.DrainingNodesFinalizer {
-				changed = true
-				continue
+			err := r.createNodeConfig(ctx, currentPod)
+			if err != nil {
+				return microerror.Mask(err)
 			}
 
-			newFinalizers = append(newFinalizers, f)
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "debug", "canceling reconciliation for pod")
+
+		} else if err != nil {
+			return microerror.Mask(err)
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "found node config for the guest cluster")
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "waiting for inspection of the reconciled pod")
 		}
 
-		if changed {
-			podToDelete = currentPod
-			podToDelete.SetFinalizers(newFinalizers)
+		r.logger.LogCtx(ctx, "level", "debug", "message", "inspecting node config for the guest cluster")
+
+		if !nodeConfig.Status.HasFinalCondition() {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has no final state")
+			resourcecanceledcontext.SetCanceled(ctx)
+			finalizerskeptcontext.SetKept(ctx)
+			r.logger.LogCtx(ctx, "debug", "canceling reconciliation for pod")
+
+			return nil
 		}
-	}
 
-	r.logger.LogCtx(ctx, "debug", "looking if the pod has to be updated or to be deleted in the Kubernetes API")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "node config of guest cluster has final state")
 
-	if podToDelete != nil {
-		r.logger.LogCtx(ctx, "debug", "the pod has to be updated or to be deleted in the Kubernetes API")
+		err = r.deleteNodeConfig(ctx, nodeConfig)
+		if err != nil {
+			return microerror.Mask(err)
+		}
```

Comment:
Why not `deleteNodeConfig` when `forcePodCleanup`?

Reply:
Good shout. Done.